File: python/TitleBar.py
import sys
from PyQt5.QtCore import *
from PyQt5.QtWidgets import *
from PyQt5.QtGui import *
from default import *

class TitleBar(QWidget):
    def __init__(self, parent):
        super(TitleBar, self).__init__()
        self.win = parent
        self.InitializeWindow()

    def InitializeWindow(self):
        self.isPressed = False
        self.setFixedHeight(TITLE_BAR_HEIGHT)
        self.InitializeViews()
        pass

    def InitializeViews(self):
        self.iconLabel = QLabel(self)
        self.titleLabel = QLabel(self)

        self.minButton = QPushButton(self)
        self.restoreButton = QPushButton(self)
        self.closeButton = QPushButton(self)

        self.minButton.setFixedSize(TITLE_BUTTON_SIZE, TITLE_BUTTON_SIZE);
        self.restoreButton.setFixedSize(TITLE_BUTTON_SIZE, TITLE_BUTTON_SIZE);
        self.closeButton.setFixedSize(TITLE_BUTTON_SIZE, TITLE_BUTTON_SIZE);

        self.iconLabel.setFixedSize(TITLE_LABEL_SIZE, TITLE_LABEL_SIZE);
        self.titleLabel.setFixedHeight(TITLE_LABEL_SIZE);

        self.iconLabel.setAlignment(Qt.AlignCenter);
        self.titleLabel.setAlignment(Qt.AlignCenter);

        self.minButton.setIcon(QIcon(TITLE_MIN_ICON));
        self.restoreButton.setIcon(QIcon(TITLE_RESTORE_ICON));
        self.closeButton.setIcon(QIcon(TITLE_CLS_ICON));

        self.minButton.clicked.connect(self.ShowMininizedWindow)
        self.restoreButton.clicked.connect(self.ShowRestoreWindow)
        self.closeButton.clicked.connect(self.CloseWindow)

        self.lay = QHBoxLayout(self)
        self.setLayout(self.lay)

        self.lay.setSpacing(0)
        self.lay.setContentsMargins(0, 0, 0, 0)

        self.lay.addWidget(self.iconLabel)
        self.lay.addWidget(self.titleLabel)
        self.lay.addWidget(self.minButton)
        self.lay.addWidget(self.restoreButton)
        self.lay.addWidget(self.closeButton)

    def ShowMininizedWindow(self):
        self.win.showMinimized()

    def ShowMaximizedWindow(self):
        self.window.showMaximized()

    def ShowRestoreWindow(self):
        if self.win.isMaximized():
            self.win.showNormal()
        else:
            self.win.showMaximized()

    def CloseWindow(self):
        self.win.close()

    def SetTitle(self, str):
        self.titleLabel.setText(str)

    def SetIcon(self, pix):
        self.iconLabel.setPixmap(pix.scaled(self.iconLabel.size() - QSize(TITLE_ICON_MAG, TITLE_ICON_MAG)))

    def mouseDoubleClickEvent(self, event):
        self.ShowRestoreWindow()
        return QWidget().mouseDoubleClickEvent(event)

    def mousePressEvent(self, event):
        self.isPressed = True
        self.startPos = event.globalPos()
        return QWidget().mousePressEvent(event)

    def mouseReleaseEvent(self, event):
        self.isPressed = False
        return QWidget().mouseReleaseEvent(event)

    def mouseMoveEvent(self, event):
        if self.isPressed:
            if self.win.isMaximized:
                self.win.showNormal()

            movePos = event.globalPos() - self.startPos
            self.startPos = event.globalPos()
            self.win.move(self.win.pos() + movePos)

        return QWidget().mouseMoveEvent(event)

if __name__ == '__main__':
    app = QApplication(sys.argv)
    win = TitleBar(None)
    win.show()
    sys.exit(app.exec_())
    pass

File: cpp/CustomTitleBar/TitleBar.cpp
#include "TitleBar.h"

TitleBar::TitleBar(QWidget *parent /* = nullptr */)
	:QWidget(parent)
{
	this->win = parent;
	InitializeWidget();
}

TitleBar::~TitleBar()
{
}

void TitleBar::InitializeWidget()
{
	isPressed = false;
	setFixedHeight(TITLE_BAR_HEIGHT);

	InitializeViews();
}

void TitleBar::InitializeViews()
{
	iconLabel = new QLabel(this);
	titleLabel = new QLabel(this);

	minButton = new QPushButton(this);
	restoreButton = new QPushButton(this);
	closeButton = new QPushButton(this);

	minButton->setFixedSize(TITLE_BUTTON_SIZE, TITLE_BUTTON_SIZE);
	restoreButton->setFixedSize(TITLE_BUTTON_SIZE, TITLE_BUTTON_SIZE);
	closeButton->setFixedSize(TITLE_BUTTON_SIZE, TITLE_BUTTON_SIZE);

	iconLabel->setFixedSize(TITLE_LABEL_SIZE, TITLE_LABEL_SIZE);
	titleLabel->setFixedHeight(TITLE_LABEL_SIZE);

	iconLabel->setAlignment(Qt::AlignCenter);
	titleLabel->setAlignment(Qt::AlignCenter);

	minButton->setIcon(QIcon(TITLE_MIN_ICON));
	restoreButton->setIcon(QIcon(TITLE_RESTORE_ICON));
	closeButton->setIcon(QIcon(TITLE_CLS_ICON));

	connect(minButton, &QPushButton::clicked, this, &TitleBar::ShowMinimizedWindow);
	connect(restoreButton, &QPushButton::clicked, this, &TitleBar::ShowRestoreWindow);
	connect(closeButton, &QPushButton::clicked, this, &TitleBar::CloseWindow);

	lay = new QHBoxLayout(this);
	this->setLayout(lay);
	lay->setSpacing(0);
	lay->setMargin(0);

	lay->addWidget(iconLabel);
	lay->addWidget(titleLabel);
	lay->addWidget(minButton);
	lay->addWidget(restoreButton);
	lay->addWidget(closeButton);
}

void TitleBar::ShowMaximizedWindow()
{
	win->showMaximized();
}

void TitleBar::ShowMinimizedWindow()
{
	win->showMinimized();
}

void TitleBar::ShowRestoreWindow()
{
	if (win->isMaximized())
	{
		win->showNormal();
	}
	else
	{
		win->showMaximized();
	}
}

void TitleBar::CloseWindow()
{
	win->close();
}

void TitleBar::SetTitle(const QString &str)
{
	titleLabel->setText(str);
}

void TitleBar::SetIcon(QPixmap &pix)
{
	iconLabel->setPixmap(pix.scaled(iconLabel->size() - QSize(TITLE_ICON_MAG, TITLE_ICON_MAG)));
}

void TitleBar::mouseDoubleClickEvent(QMouseEvent *event)
{
	ShowRestoreWindow();
	
	return QWidget::mouseDoubleClickEvent(event);
}

void TitleBar::mousePressEvent(QMouseEvent *event)
{
	/*if (win->isMaximized())
		return;*/

	isPressed = true;
	startPos = event->globalPos();

	return QWidget::mousePressEvent(event);
}

void TitleBar::mouseReleaseEvent(QMouseEvent *event)
{
	isPressed = false;

	return QWidget::mouseReleaseEvent(event);
}

void TitleBar::mouseMoveEvent(QMouseEvent *event)
{
	if (isPressed)
	{
		if (win->isMaximized())
		{
			win->showNormal();
		}

		QPoint movePos = event->globalPos() - startPos;
		startPos = event->globalPos();

		win->move(win->pos() + movePos);
	}

}